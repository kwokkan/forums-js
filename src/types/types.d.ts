
interface IForumSessionUser {
    id: number;
    name: string;
    joinedDate: number;
}

interface INextSession extends any {
    forumsUser?: IForumSessionUser;
}

interface INextProvider extends any {
    id: string;
    name: string;
}

declare module "next-auth" {
    interface INextAccount extends any {
        provider: string;
        id: string;
    }

    export interface IProviderOptions {
        providers: any[];
        callbacks?: IProviderOptionsCallback;
        events?: INextEventsOptions;
        pages?: INextPageOptions;
        debug?: boolean;
    }

    interface IProviderOptionsCallback {
        signIn?: (user: INextUser, account: INextAccount, profile: INextProfile) => Promise<bool>;
        redirect?: (url: string, baseUrl: string) => Promise<string>;
        session?: (session: INextSession, user: INextUser, sessionToken: string) => Promise<INextSession>;
        jwt?: (token: INextToken, user: INextUser, account: INextAccount, profile: INextProfile, isNewUser: boolean) => Promise<INextToken>;
    }

    interface INextEventsOptions {
        signIn?: (message: any) => Promise;
        signOut?: (message: any) => Promise;
        createUser?: (message: any) => Promise;
        linkAccount?: (message: any) => Promise;
        session?: (message: any) => Promise;
        error?: (message: any) => Promise;
    }

    interface INextPageOptions {
        signIn?: string;
        signOut?: string;
        error?: string;
        verifyRequest?: string;
        newUser?: string;
    }

    interface INextProfile extends any {
    }


    interface INextToken extends any {
        forumsUser?: IForumSessionUser;
    }

    interface INextUser extends any {
        name: string;
        forumsUser?: IForumSessionUser;
    }

    function handler(req: IncomingMessage, res: ServerResponse, options: IProviderOptions): void;

    export = handler;
}

declare module "next-auth/client" {
    export const Provider: any;

    export function getSession(context: { req: IncomingMessage }): Promise<INextSession>;

    export function providers<T>(context: GetServerSidePropsContext<T>): Promise<INextProvider[]>;

    export function signIn(provider?: string, args?: { callbackUrl?: string }): void;

    export function signOut(): void;

    export function useSession(): [INextSession | null, boolean];
}

declare module "next-auth/providers" {
    interface IProviderConfig {
        clientId: string,
        clientSecret?: string,
        scope?: string[] | string
    }

    export = {
        GitHub: (config: IProviderConfig) => any
    };
}
