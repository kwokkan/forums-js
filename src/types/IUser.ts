export interface IUser {
    id: number;
    name: string;
    joinedDate: Date;
}
