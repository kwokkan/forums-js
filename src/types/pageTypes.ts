import { GetServerSideProps } from "next";
import { ParsedUrlQuery } from "querystring";

export type GetTypedServerSideProps<TParams extends ParsedUrlQuery, TProps = { [key: string]: any }> = GetServerSideProps<TProps, TParams>;
