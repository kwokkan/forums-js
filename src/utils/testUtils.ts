import Enzyme from "enzyme";
import Adapter from "@wojtekmaj/enzyme-adapter-react-17";

export function setupEnzyme() {
    Enzyme.configure({ adapter: new Adapter() });
}
