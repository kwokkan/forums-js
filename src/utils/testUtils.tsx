import React from "react";
import { findDOMNode } from "react-dom";
import { renderIntoDocument } from 'react-dom/test-utils';

class TestWrapper extends React.Component {
    render() {
        return this.props.children;
    }
}

export function getInstance(children: React.ReactElement<any, any>): JSX.Element {
    // isReactComponent is only defined if children is of React.Component class
    // so we can test against this to verify this is a functional component
    if (!(children.type.prototype && children.type.prototype.isReactComponent)) {
        const instanceRef = React.createRef();

        const Forwarded = React.forwardRef((props, ref) => React.cloneElement(children, { props: props, ref: ref }));

         renderIntoDocument(
            <TestWrapper><Forwarded  /></TestWrapper>
        ) as unknown as React.ReactElement;

        return instanceRef.current as unknown as React.ReactElement;
    }

    return renderIntoDocument(children) as unknown as React.ReactElement;
}

export function getDOMNode(children: React.ReactElement<any, any>): JSX.Element {
    return findDOMNode(getInstance(children) as unknown as React.Component) as unknown as JSX.Element;
}
