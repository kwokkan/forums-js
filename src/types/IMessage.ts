import { IUser } from "./IUser";

export interface IMessage {
    id: number;
    created: Date;
    content: string;
    user: IUser;
}
