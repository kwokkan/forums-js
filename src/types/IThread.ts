import { IMessage } from "./IMessage";

export interface IThread {
    id: number;
    created: number;
    name: string;
    messages: IMessage[];
}
