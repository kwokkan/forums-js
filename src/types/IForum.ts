import { IThread } from "./IThread";

export interface IForum {
    id: number;
    created: Date;
    name: string;
    description?: string;
    threads?: IThread[];
}
