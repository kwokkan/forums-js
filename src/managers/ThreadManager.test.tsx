/**
 * @jest-environment jsdom
*/
import { mount } from "enzyme";
import { act } from "react-dom/test-utils";
import { create } from "react-test-renderer";
import { IMessage } from "../types/IMessage";
import { IThread } from "../types/IThread";
import { IUser } from "../types/IUser";
import { setupEnzyme } from "../utils/testUtils";

beforeAll(() => {
    setupEnzyme();
});

test("Renders without error", () => {
    jest.isolateModules(() => {
        const thread: IThread = {
            created: 1,
            id: 2,
            forumId: 1,
            messages: [],
            name: "Test thread"
        };

        const { ThreadManager } = require("./ThreadManager");

        const tree = create(
            <ThreadManager thread={thread} breadcrumbs={[]} />
        ).toJSON();

        expect(tree).toMatchSnapshot();
    });
});

test("onNewMessage callback called", (doneCallback) => {
    jest.isolateModules(async () => {
        const user: IUser = {
            id: 3,
            joinedDate: new Date(1900, 1, 1).getTime(),
            name: "Test user"
        };
        const mockAddMessage = jest.fn((): IMessage => ({
            content: "New message",
            created: 1,
            id: 1,
            user: user
        }));

        jest.doMock("../api", () => {
            return {
                __esModule: true,
                addMessage: mockAddMessage
            };
        });

        const thread: IThread = {
            id: 1,
            created: new Date(2020, 1, 1).getTime(),
            forumId: 1,
            name: "Test thread",
            messages: [
                {
                    id: 2,
                    created: new Date(2020, 2, 2).getTime(),
                    content: "First !!!",
                    user: user
                }
            ]
        };

        const { ThreadManager } = require("./ThreadManager");

        const wrapper = mount(
            <ThreadManager thread={thread} breadcrumbs={[]} user={user} />
        );

        act(() => {
            wrapper.find("textarea").simulate("change", { target: { value: "New message" } });
        });

        await act(async () => {
            wrapper.find("button").simulate("click");
        });

        setImmediate(() => {
            expect(mockAddMessage).toBeCalledWith(1, "New message");

            doneCallback();
        });
    });
});
