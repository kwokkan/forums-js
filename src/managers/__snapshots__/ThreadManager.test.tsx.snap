// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Renders without error 1`] = `
<div
  className="rs-panel rs-panel-default"
>
  <div
    className="rs-panel-heading"
    onClick={[Function]}
    role="rowheader"
    tabIndex={-1}
  >
    <ol
      aria-label="breadcrumbs"
      className="rs-breadcrumb"
      role="navigation"
    >
      <li
        className="rs-breadcrumb-item"
      >
        <a
          href="/"
          onClick={[Function]}
          passHref={true}
          title="Home"
        >
          <i
            className="rs-icon rs-icon-home"
          />
        </a>
      </li>
    </ol>
  </div>
  <div
    className="rs-panel-body"
  >
    <div
      className="rs-panel-group"
    />
  </div>
</div>
`;
